src-kmerge/KMerge/LoserTree.hs

```diff
@@ -35,69 +36,65 @@ placeholder = unsafeCoerce ()
 --
 -- The second half of a pair is the winner value (only losers are stored in the tree).
 --
-newLoserTree :: forall a m. (PrimMonad m, Ord a) => [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-newLoserTree [] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Nothing)
-newLoserTree [x] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Just x)
-newLoserTree xs0 = do
+newLoserTree :: forall a m. (PrimMonad m, Ord a) => NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+newLoserTree input@(a :| as) = do
```

Comment:
Not sure why you changed the names here (vs. `xs0@(x :| xs)`). We still use `x` in other functions.

You could even just do `(x :| xs)` with `!len = length xs` and then just use `len` instead of `len - 1` everywhere.

Reply:
I was trying to prevent name shadowing in `loop` and `sift`.

Reply:
Ah, yes. Then `(x0 :| xs0)` would have preserved the spirit of the original best probably (the `@` is not really needed, as we can call loop with `(x0 :| xs0)`, which works well with inlining. But I don't really mind either way.